[Streams] Support processing complete simulation and single data source

tonyghiani opened the pull request:
## 📓 Summary

Closes #241464

This work makes major changes to the processing page to rethink how we run processing on the simulation samples.
The principal changes are:
- The simulation now runs only on a single selected data source. This simplifies a lot the UX for the user and makes the simulation more granular.
- With the single data source simulation, we can differentiate what kind of simulation mode should run. So far, we were running the simulation only on the newest processors created on the current user session and not persisted yet.
Now, every data source type comes with a related simulation mode, where the **Custom samples** one is the first to enable the full pipeline simulation. This will also allow an easier integration of the failure store and original samples data sources.
- Random samples have never really been a random selection, but we were retrieving the latest 100 indexed documents. This is now renamed **Latest docs**.
- Remove data source avatar from table

Other notable changes:
- Improving the data sources management, this work introduces a persistence mechanism for the custom samples relying on the local storage. Upon refresh, the custom sample data sources won't be lost and can be correctly restored.
- The above change enables better portability of documents from external apps. The parsing CTA from Discover now works correctly in both Classic and ES|QL mode, adding the document to the processing data sources as a Custom sample source.
- Removed the deprecated `RedirectAppLinks`usages.

tonyghiani committed:
refactor(streams): switch to single selection data source

tonyghiani committed:
fix(streams): default latest samples restore

tonyghiani committed:
refactor(streams): handle data source delete

tonyghiani committed:
refactor(streams): remove data source avatar

tonyghiani committed:
refactor(streams): handle data source selection

tonyghiani committed:
fix(streams): resizing issue due to table missing overflow

tonyghiani committed:
refactor(streams): simplify data source machine

tonyghiani committed:
refactor(streams): wip complete simulation mode

tonyghiani committed:
Merge branch 'main' into 241464-update-processing-data-sources

tonyghiani committed:
refactor(streams): mvp complete simulation

tonyghiani committed:
refactor(streams): complete simulation

tonyghiani committed:
refactor(streams): minor udpates

tonyghiani committed:
refactor(streams): update simulation events

tonyghiani committed:
refactor(streams): remove deprecated RedirectAppLinks

tonyghiani committed:
refactor(streams): support both classic and ES|QL docs

tonyghiani committed:
refactor(streams): optimize name change

tonyghiani committed:
Merge branch 'main' into 241464-update-processing-data-sources

tonyghiani committed:
refactor(streams): fix merge conflicts

tonyghiani committed:
refactor(streams): minor improvements

tonyghiani committed:
refactor(streams): remove custom samples callout

tonyghiani committed:
refactor(streams): minor adjustments

tonyghiani committed:
refactor(streams): fix typing

tonyghiani commented on x-pack/platform/plugins/shared/streams_app/common/locators/streams_locator.ts:
```diff
@@ -49,10 +53,18 @@ export class StreamsAppLocatorDefinition implements LocatorDefinition<StreamsApp
+            // For custom samples data source, we need to persist the documents in the browser local storage to retrieve them later.
+            if (dataSource.type === 'custom-samples') {
+              return parseAndPersistCustomSamplesDataSource(dataSource, params.name);
+            }
```
**note**: to achieve portability of custom samples from other apps, like Discover in this case, the docs are persisted on the local storage by a specified storage key that is shared by the url state and used to restore the custom samples.

tonyghiani committed:
test(streams): adapt tests

tonyghiani committed:
Merge branch 'main' into 241464-update-processing-data-sources

kibanamachine committed:
Changes from node scripts/lint_ts_projects --fix

tonyghiani committed:
fix(streams): fix broken loading state

tonyghiani committed:
Merge branch '241464-update-processing-data-sources' of github.com:tonyghiani/kibana into 241464-update-processing-data-sources

jughosta reviewed (approved):
Data Discovery changes LGTM 👍 
Code review only.

tonyghiani committed:
fix(fields_metadata): prevent bundle explosion and update limits

jbudz reviewed (approved):
`packages/kbn-optimizer/limits.yml` LGTM

tonyghiani committed:
Merge branch 'main' into 241464-update-processing-data-sources

tonyghiani committed:
fix(fields_metadata): merge conflicts

tonyghiani commented on x-pack/platform/plugins/shared/streams_app/common/locators/streams_locator.ts:
```diff
@@ -67,3 +79,24 @@ export class StreamsAppLocatorDefinition implements LocatorDefinition<StreamsApp
+  dataSource: CustomSamplesDataSource,
+  streamName: string
+) => {
+  const key = dataSource.storageKey ?? `streams:${streamName}__${uuidv4()}`;
```
It's a fair concern, in the case of pulling a doc from Discover it would just store one document for every visited stream. In case the user makes the same flows on different documents from the same stream, only the last one is stored on the localStorage and overrides the previous one. Is that enough as a tradeoff? I liked the idea from a user perspective to jump from different streams and have a sampled document available if they keep the stream page open.

Another option is to rely on the sessionStorage instead of localStorage, which would automatically delete the stored document when closing the page.

tonyghiani committed:
fix(streams): revert locator type change

tonyghiani committed:
Merge branch 'main' into 241464-update-processing-data-sources

tonyghiani committed:
refactor(streams): remember persisted custom samples data sources

tonyghiani committed:
t push
:wq
:wqMerge branch '241464-update-processing-data-sources' of github.com:tonyghiani/kibana into 241464-update-processing-data-sources

tonyghiani committed:
Merge branch 'main' into 241464-update-processing-data-sources

tonyghiani commented:
> To be clear, you are suggesting that this could be done, but that's not the case now though, right? It keeps forgetting about the existing custom data sources:
> <img width="696" height="108" alt="Screenshot 2025-11-10 at 11 36 22" src="https://github.com/user-attachments/assets/3334fdb3-c620-49c7-b07c-b9f337cdc2a7" />

These listed in the screenshot are not stored on the Discover flow of pulling a document into the stream, those are repeatedly generated (and not removed) when you manually add a custom samples data source. For the discover document, there is always the same key used for each stream, so it doesn't pollute the localstorage if a user keeps pulling docs from discover, they'll get the last one used.
I did a bit of extra mile and now works better:
- Remove storage entry when the data source is manually removed.
- Restore custom samples sources when the user navigates between pages (not applied to kql, which still relies on the url)
Doing so, users won't lose their custom samples even while navigating across different streams, and there is no pollution because only the users' custom sample data sources are stored.

tonyghiani committed:
refactor(streams): fix types

tonyghiani commented:
> Why not having custom sample datasources, persisting them to local or session storage and loading them from there?

Not sure I get this question Joe, happy to also chat offline. Do you mean having custom samples data sources being shared and available to any stream processing page a data source? If so, why would we show a data source of documents for a stream that has nothing to do with the current stream under processing? If that's not the point, let's chat offline to get to the point fast 👌

tonyghiani committed:
refactor(streams): get all the persisted custom samples from sessionStorage

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/359916)
* Commit: df09d6a1aa8b40fab4faac39ed650c65e94b7078

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/359916#019a6e99-cb0e-42d3-ac81-c31a56e89baa)

### History
* :green_heart: [Build #359859](https://buildkite.com/elastic/kibana-pull-request/builds/359859) succeeded 7062946531221e721d0eab6cf3959e7c8146d220
* :broken_heart: [Build #359822](https://buildkite.com/elastic/kibana-pull-request/builds/359822) failed b9ab7ddf77122616ecc1a09f59b3dc437f422d58
* :broken_heart: [Build #359651](https://buildkite.com/elastic/kibana-pull-request/builds/359651) failed 6155e6961c8d0f6f0778b81ecf6f3e64e98826c7
* :green_heart: [Build #359477](https://buildkite.com/elastic/kibana-pull-request/builds/359477) succeeded 20e60a1a6a361da1f8e91a92ca762299c7c02287
* :yellow_heart: [Build #358926](https://buildkite.com/elastic/kibana-pull-request/builds/358926) was flaky c1f9852639264acf52b320d85ebaf392bcd06df2
* :broken_heart: [Build #358902](https://buildkite.com/elastic/kibana-pull-request/builds/358902) failed 8eddb2b466cefbba7022db332168ca537916b12e

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359916","number":359916,"commit":"df09d6a1aa8b40fab4faac39ed650c65e94b7078"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359859","number":359859,"commit":"7062946531221e721d0eab6cf3959e7c8146d220"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359822","number":359822,"commit":"b9ab7ddf77122616ecc1a09f59b3dc437f422d58"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359651","number":359651,"commit":"6155e6961c8d0f6f0778b81ecf6f3e64e98826c7"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359477","number":359477,"commit":"20e60a1a6a361da1f8e91a92ca762299c7c02287"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/358926","number":358926,"commit":"c1f9852639264acf52b320d85ebaf392bcd06df2"}],"number":359916}
buildkite-pr-comment-->

tonyghiani committed:
refactor(streams): i18n name

Kerry350 reviewed (commented):
I'm not sure what was decided between these two comments offline (https://github.com/elastic/kibana/pull/241809#issuecomment-3512016861 and https://github.com/elastic/kibana/pull/241809#issuecomment-3512058018) but this otherwise looks good to me 👍

tonyghiani committed:
Update x-pack/platform/plugins/shared/streams_app/public/components/data_management/stream_detail_enrichment/data_sources_flyout/translations.tsx

Co-authored-by: Kerry Gallagher <k.gallagher.05@gmail.com>

tonyghiani commented:
> I'm not sure what was decided between these two comments offline

Hey Kerry, we discussed that offline and the agreement we had is implemented with [refactor(streams): get all the persisted custom samples from sessionS…](https://github.com/elastic/kibana/pull/241809/commits/df09d6a1aa8b40fab4faac39ed650c65e94b7078)

tonyghiani committed:
Merge branch 'main' into 241464-update-processing-data-sources

tonyghiani committed:
feat(streams): add simulation badge in selector

tonyghiani committed:
refactor(streams): get raw docs always

tonyghiani commented:
> You need to open the data source flyout to see whether it's a "new processing" or "full pipeline" simulation mode. I think it makes sense to show that next to the picker somehow (e.g. when you come from discover, you never see that) - wdyt about showing the badge there as well?

It makes sense, we didn't address all the design concerns here to unblock this work, but this was a quick addition, I added it with [feat(streams): add simulation badge in selector](https://github.com/elastic/kibana/pull/241809/commits/d0d4cce07d2410eae1a2915f04f903a706448200)

> Not sure whether it broke with this PR or was broken before, but when using the Discover data we pick up the formatted date field, which breaks

It was broken with this PR, but i noticed something while fixing this: in Discover for the JSON view the document is fetched from ES, so we cannot port it in the right format relying on the custom samples.
I think we need to differentiate how the document is pulled depending on the mode.
I changed it such that:
- When the user is in Classic mode, the document is pulled with a KQL source, as it used to work so far
- When the user is in ES|QL mode, the document is pulled with a custom sample source.

flash1293 reviewed (commented):
Thanks @tonyghiani , I like the badge!

> When the user is in Classic mode, the document is pulled with a KQL source, as it used to work so far
> When the user is in ES|QL mode, the document is pulled with a custom sample source.

This makes sense to me, but now one of them is processed end-to-end and the other isn't - this is a bit weird. I'm only thinking about it right now, but I guess in most cases the samples coming from Discover can't reliably be simulated end-to-end, right?

Not sure what the right behavior even should be here... if @LucaWintergerst doesn't have concerns I'm OK to go with this for now

tonyghiani committed:
refactor(streams): fix tests

tonyghiani committed:
Merge branch 'main' into 241464-update-processing-data-sources

flash1293 reviewed (approved):
>  I can't honestly think of a better solution if we want to pull the document from the Discover 🤔

We could have a "from discover" data source which is like custom-source but only simulates partial processing. But that sounds like overcomplicating the whole thing, let's not do that.

>  I'd proceed with this, which at least let the user get a document correctly, and discuss afterwards how we should simulate.

Agreed from my side if @LucaWintergerst is OK with it.

Btw, looking even deeper, the ESQL-transfer from Discover is also not working in other ways:
<img width="1421" height="587" alt="Screenshot 2025-11-12 at 14 52 20" src="https://github.com/user-attachments/assets/ecdbb211-2aec-4427-85eb-d110b420507a" />

ESQL doesn't know about aliases vs. _real_ fields, so we get all of them, but we can't write the aliases back... Maybe we should just remove this for ESQL for now and simply bring the user to the "Latest samples" data source.

Not blocking for this PR since it's pre-existing

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/360553)
* Commit: ead83e08fb05cb3208512f80b60b18203078155b

### Failed CI Steps
* [Linting](https://buildkite.com/elastic/kibana-pull-request/builds/360553#019a7849-b414-459e-8dcf-e58626f86962)

### History
* :broken_heart: [Build #360474](https://buildkite.com/elastic/kibana-pull-request/builds/360474) failed 9174edf861103e78beabf7996c34e64ab74d15f8
* :broken_heart: [Build #359916](https://buildkite.com/elastic/kibana-pull-request/builds/359916) failed df09d6a1aa8b40fab4faac39ed650c65e94b7078
* :green_heart: [Build #359859](https://buildkite.com/elastic/kibana-pull-request/builds/359859) succeeded 7062946531221e721d0eab6cf3959e7c8146d220
* :broken_heart: [Build #359822](https://buildkite.com/elastic/kibana-pull-request/builds/359822) failed b9ab7ddf77122616ecc1a09f59b3dc437f422d58
* :broken_heart: [Build #359651](https://buildkite.com/elastic/kibana-pull-request/builds/359651) failed 6155e6961c8d0f6f0778b81ecf6f3e64e98826c7
* :green_heart: [Build #359477](https://buildkite.com/elastic/kibana-pull-request/builds/359477) succeeded 20e60a1a6a361da1f8e91a92ca762299c7c02287

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/360553","number":360553,"commit":"ead83e08fb05cb3208512f80b60b18203078155b"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/360474","number":360474,"commit":"9174edf861103e78beabf7996c34e64ab74d15f8"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359916","number":359916,"commit":"df09d6a1aa8b40fab4faac39ed650c65e94b7078"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359859","number":359859,"commit":"7062946531221e721d0eab6cf3959e7c8146d220"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359822","number":359822,"commit":"b9ab7ddf77122616ecc1a09f59b3dc437f422d58"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359651","number":359651,"commit":"6155e6961c8d0f6f0778b81ecf6f3e64e98826c7"}],"number":360553}
buildkite-pr-comment-->

tonyghiani committed:
fix(streams): remove 'only' from data sources management test

tonyghiani committed:
Merge branch 'main' into 241464-update-processing-data-sources
